### DISCUSSION TITLE
Autok3s Air-gap installation design

### DISCUSSION THREAD

**Discussion by orangedeng** [cnrancher/autok3s#480] [Ideas]
# autok3s air-gap installation

### storage backend support(not necessary)

Maybe we can support multiple storage backend for storing air-gap package.

- Local file system(**Must**)
- S3(Nice to have)

If we don't need to management storage backend, we only need to support local file system to store packages.

### air-gap package management

We need to management air-gap package by k3s version. Archs will be selected when generating/updating package. Following archs will be supported:

- amd64
- arm64
- armhf
- s390x(recently supported by k3s)

Following resources will be downloaded for each:

- binary(k3s itself)
- offline images by arch
- image list
- checksum file

Those resource will be stored in following directory:

```bash
.
├── amd64
│   ├── k3s
│   ├── k3s-airgap-images-amd64.tar.gz
│   ├── k3s-images.txt
│   └── sha256sum-amd64.txt
└── version.json
```

The `version.json` will store k3s version and the target version info.

The package also support to be **imported**/**exported** in tar.gz format. The autok3s will check the air-gap package before cluster setup.

Following two functions will also be needed.

1. The k3s `install.sh` script should be downloaded when compiling new autok3s release and stored in database. The script's source and version are needed to be recorded.
2. The air-package binaries download site should support being configured as then binaries can be found in [Github](https://github.com/k3s-io/k3s/releases) and [Aliyun OSS](https://rancher-mirror.oss-cn-beijing.aliyuncs.com).

### air-gap cluster installation

For creating new cluster, following option(s) will be added:

- Air-gap package path

And the current options will not be compatible with the new air-gap option(s):

- K3s channel
- K3s version
- K3s install script
- System Default Registry

If the `Air-gap package path` is specified, autok3s will assume that the cluster is using air-gap package to install.  
When the cluster installation is using air-gap package, the related package and install script will be scp to cluster node before running `install.sh` and those resource will be un-tar and moved to target path.

#### air-gap CLI

Add `airgap` sub-command, following functions will be needed:

- ls
- create
- update
- rm
- import
- export
- update-install-script

The `update-install-script` command is for udpating the stored `install.sh` script. The auto-update mode will be executed if no flags are specified. Or specifying the `file` flag to update the stored install script.

#### air-gap UI

There will be a new page `Air-gap Packages` under `Settings` and list all the air-gap packages with following columns:

- name
- k3s version
- included archs

In the list page, we will have `Show/Update Install Script`, `Show/Update Package Source` and `Create` button.

The `Show/Update Install Script` will be a **popup-form**. And the current version(commit) and the [source](https://github.com/kingsd041/k3s/blob/master/install.sh) of the script will be **readonly** in page. And it will have a **multi-line input** for displaying and inputing **the install script**.  
The `Show/Update Package Source` will be a **popup-form**. And the Air-gap package source will display/input with a **single line text input**.

The `Create` will be the a normal page. The name input will be required and the `import` options will let user to select a current tar.gz file path to upload. If the `import` is specified, No other options will be required. Backend will take the file upload first(with api `/v1/airgaps?action=upload`) and response the `filepath`, `k3s version` and the `included archs` to show in UI.  
Otherwise, the `k3s version` and `archs` will be required for air-gap package create.

The actions for each package are following:

- update
- remove
- export

The `update` action will display the **multiple selection checkbox** for selecting **archs** with `name` and `k3s version` being **readonly**.  
The `export` action will let user to select a path to store the export tar.gz file. Kind of like a download function.

#### cluster create CLI

There will be a air-gap package path(`air-gap-package`) flags will be added to cluster create options. Refer to the design of [air-gap install design](#air-gap-cluster-installation), some flags will not be compatible with the air-gap flag and will be checked in cluster create process.  
And this tag should be input as a file path with the `path` column of `air-gap ls` output.

#### cluster create UI

In the `K3s Options` section of cluster create page, The air-gap install will be a **checkbox** and the air-gap package list **dropdown with single selection**. If the air-gap install is checked, the `K3s Channel`, `K3s Version`, `K3s Install Script` and `System Default Registry` will be hidden.

### air-gap options for cluster upgrade

If the cluster used to install with air-gap installaion, the upgrade cluster page will be set to select air-gap package page instead of inputing `K3s Channel`, `K3s Version` and `K3s Install Script` and the air-gap install checkbox will be checked state.  
It is supported to un-checked the air-gap install options to use online install when upgrading a cluster.

**Comment by codering** [comment]
Great， I like it.

**Comment by orangedeng** [comment]
PR for airgap support: https://github.com/cnrancher/autok3s/pull/491

**Comment by orangedeng** [comment]
The UI function `Show/Update Package Source` will be supported in the feature.
This feature has been implemented in v0.8.0
